### ISSUE TITLE
`colorsys` returning out of bounds values

### ISSUE THREAD

**Issue by FFY00** [python/cpython#102581]
# Bug report

On certain inputs, the `colorsys` module returns values that are out of the documented allowed range, see GH-100754 for the previous discussion.

# Your environment

- CPython versions tested on: 3.12
- Operating system and architecture: any

### Linked PRs
* gh-106254

**Comment by FFY00** [comment]
I think the first thing to do here is to validate that our conversion logic is actually correct. If it is, then that means that for the given input value, there's no valid output.

In such case, I'd propose we raise a warning when the result goes out of bounds, and add a new parameter to the function(s) to clip the value, as that's what most users will want top do. Opting in to the clipping behavior makes sure we don't break anything, and still provides a reasonable solution for users that are okay with a result approximation.

I'm not sure if I am being too defensive here or not :sweat_smile:

**Comment by FFY00** [comment]
@KelSolaar, you maintain https://github.com/colour-science (which is awesome!), do you have any input regarding this issue?

I am not a color theory expert by any means, so I am really not equipped to analyze the technical part of this.

**Comment by ppwwyyxx** [comment]
Agree with @tacaswell that we should:
(1) Warn or raise error if input is out of bound
(2) Unconditionally clip the output.

(2) is reasonable because, we can mathematically prove that, the output must be in bound as long as the input is in bound. Since callers are responsible for providing in-bound inputs (checked in (1)), the only reason why outputs may be out of bound is the limited floating point precision. In that case clipping will simply make the results more correct.

Here is a test case I manually OCRed from #100754:

```
import colorsys
r = 0.6439640003843092
g = 0.2082993986149012
b = 1.0
print(colorsys.rgb_to_hls(r, g, b))
(0.7583815988449848, 0.6041496993074507, 1.0000000000000002)
```

**Comment by KelSolaar** [comment]
Hi,

@FFY00: Thanks for dragging me in!

There is no good reason why it should be systematically clipped. RGB values are either:

1.  Bounded, typically for standard dynamic range, i.e. SDR 8, 10, 12 and 16-bit encodings with minimum representable value of 0 and maximum representable values of 255, 1023, 4095, 65535 respectively. They are also encoded with a non-linear transfer function, e.g. power (gamma) or log. Once decoded and normalised to floating point representation, they will be in range [0. 1].
2.  Unbounded, typically for high dynamic range, i.e. HDR 16-half, 32-bit, 64-bit and 128-bit float. The minimum and maximum representable values are more complicated to compute, e.g. [-65504, 65504] range for 16-half. They are encoded linearly and can represent either relative of absolute values.

With that in mind, in the film industry where I work, we pretty much exclusively work with the second case and HDR imagery thus the following behaviour is totally expected and desirable to us:

```python
>>> import colour
>>> colour.RGB_to_HSL([1500, 2500, 3000])
array([  5.55555556e-01,  -3.33481547e-01,   2.25000000e+03])
>>> colour.HSL_to_RGB(colour.RGB_to_HSL([1500, 2500, 3000]))
array([ 1500.,  2500.,  3000.])
```

As a general rule, we never clip in Colour unless it is absolutely required.

Hope that helps!

Cheers,

Thomas

**Comment by tacaswell** [comment]
Given @KelSolaar comment, I would argue that either the docs should be updated to indicate that these functions supports case 2 and drop the [0, 1] comments from the docs _or_ the docs should make clear these functions are only meant to support case 1 and then clip.

It is my suspicion that most users who reach for `colorsys` (rather than a more specialized tool) are working in case 1, but I have no actual data one way or the other.

**Comment by KelSolaar** [comment]
> It is my suspicion that most users who reach for colorsys (rather than a more specialized tool) are working in case 1, but I have no actual data one way or the other.

This is certainly true today, but it might not be tomorrow as HDR adoption grows, maybe a way to cut the apple in two is to change the function signature and add a `clip` argument that defaults to True and explain some of the reasoning, e.g. if you work with HDR then you can set to `False`. Note the you will probably want to “sanitize” and clip HSV/HSL value/ligthness input also for consistency.

**Comment by MonadChains** [comment]
Honestly I feel that the entire module is inconsistent. For example in the case rgb -> yiq there is no clipping:
https://github.com/python/cpython/blob/6b5166fb12c4744544da4ee26ef437d025eb762a/Lib/colorsys.py#L40-L44
while viceversa there is:
https://github.com/python/cpython/blob/6b5166fb12c4744544da4ee26ef437d025eb762a/Lib/colorsys.py#L46-L67
I would propose to uniform it either by making all functions working for unlimited color space with the option to clip to remain in the space [0,1] or agree to a unique way of handling overflowing values (throwing a warning or clipping).

@KelSolaar in the case of unbounded RGB values, are the conversions to the other representations available in the module (YIQ, HLS, HSV) reversible (i.e. applying the inverse transformation should return the initial unbounded value)?

**Comment by KelSolaar** [comment]
Hello,

> I would propose to uniform it either by making all functions working for unlimited color space with the option to clip to remain in the space [0,1] or agree to a unique way of handling overflowing values (throwing a warning or clipping).

I would certainly agree with that.

> @KelSolaar in the case of unbounded RGB values, are the conversions to the other representations available in the module (YIQ, HLS, HSV) reversible (i.e. applying the inverse transformation should return the initial unbounded value)?

I don't really understand why `rgb_to_yiq` does not clip why the inverse path does not. All the 3 transforms should be invertible and the unit tests should try to test the roundtrip ideally if not already.

**Comment by ghost** [comment]
Hi, is anybody working on this , if not any suggestions?

**Comment by tacaswell** [comment]
https://github.com/python/cpython/issues/102581#issuecomment-1470558199  Is likely the best option but if that is an unacceptable expansion of the API, then I think clipping and documenting that for HDR cases users need to look for other tools is the second best choice.  Either way the clipping (or not) should be consistent across all of the functions.

**Comment by ghost** [comment]
> [#102581 (comment)](https://github.com/python/cpython/issues/102581#issuecomment-1470558199) Is likely the best option but if that is an unacceptable expansion of the API, then I think clipping and documenting that for HDR cases users need to look for other tools is the second best choice. Either way the clipping (or not) should be consistent across all of the functions.

I agree and I think that enforcing clipping in all functions and documenting is the ideal choice. I can open a PR if it's not already handled ?
